### ISSUE TITLE
Spider: ​​​​​Illinois Health Facilities and Services Review Board

### ISSUE THREAD

**Issue by pjsier** [City-Bureau/city-scrapers#1001]
URL: https://www2.illinois.gov/sites/hfsrb/events/Pages/Board-Meetings.aspx
Spider Name: il_health_facilities
Agency Name: ​​​​​Illinois Health Facilities and Services Review Board

**Comment by masoodqq** [comment]
I would like to work on this issu.

**Comment by pjsier** [comment]
@masoodqq sounds great! Assigning you now

**Comment by Ni3dzwi3dz** [comment]
Hi,
is this issue still opened? If yes, i`m willing to help.

**Comment by palakshivlani-11** [comment]
Hi,
is this issue still open

**Comment by haileyhoyat** [comment]
@palakshivlani-11 Hi. Thanks so much for checking out the project. Go for it.

**Comment by yawar1101** [comment]
willing to help

**Comment by haileyhoyat** [comment]
@yawar1101 hi. thanks so much for checking us out. go for it.

**Comment by godclause** [comment]
Hi @haileyhoyat!

I'd like to try my hand on this one.

**Comment by haileyhoyat** [comment]
@godclause Hello! Go for it. Cheers.

**Comment by godclause** [comment]
@haileyhoyat Hi!  Is this, Github, the most currently accessible venue for contributors to correspond on questions about the project?   (I did read the documentation, which is why I'm asksing...)

**Comment by godclause** [comment]
@haileyhoyat Hi! 

Is there a standard procedure / code of conduct, etc. that I should follow for applying security and dependency updates here?

**Comment by appills** [comment]
> 1. Is the general consensus around scraping PDF documents associated to an idea that they absolutely must be downloaded, stored and locally scraped?

No need to store them locally. You're going to get a stream of bytes regardless, so you can just read the byte stream directly from the request. I don't know how package management is applied here (e.g. pypdf)

```import requests
from io import BytesIO
from pypdf import PdfReader

resp = requests.get("https://hfsrb.illinois.gov/content/dam/soi/en/web/hfsrb/events/documents/2024/january-23-2024-state-board-meeting-/21-007(4)%20Permit%20Renewal%20Winchester%20ASTC.pdf")
print(resp.headers)
reader = PdfReader(BytesIO(resp.content))
print(reader.pages[0].extract_text())```

**Comment by godclause** [comment]
> > 1. Is the general consensus around scraping PDF documents associated to an idea that they absolutely must be downloaded, stored and locally scraped?
> 
> No need to store them locally. You're going to get a stream of bytes regardless, so you can just read the byte stream directly from the request. I don't know how package management is applied here (e.g. pypdf)
> 
> ```
> from io import BytesIO
> from pypdf import PdfReader
> 
> resp = requests.get("https://hfsrb.illinois.gov/content/dam/soi/en/web/hfsrb/events/documents/2024/january-23-2024-state-board-meeting-/21-007(4)%20Permit%20Renewal%20Winchester%20ASTC.pdf")
> print(resp.headers)
> reader = PdfReader(BytesIO(resp.content))
> print(reader.pages[0].extract_text())
> ```

I have a sustained perspective that running python operations is not wholly shell agnostic (e.g. zsh, bash, csh).  With this in mind, I believe package management has proven to be problematic and ought to be considered when starting a project.  

The problem is that zsh is now the default shell on macs.  The commands provided in our docs support bash scripting. 

For the folks running zsh, should there be verbiage included in our docs to remind us to consider changing our shell if we're running Catalina and beyond?

If not, should there be language updated in the docs that explains the following scenarios:

For Macs
1. Switch to bash shell and run all commands from there.
2. Build a Linux VM [virtual machine] and run commands from within that.
3. Build a Windows VM and run commands from within that.

Is any of this at all necessary?

**Comment by onyangojerry** [comment]
@godclause 

yes, there should be verbiage guiding us and or reminding us to either switch to bash or use a virtual machine just so to save us the the agony that comes with continuous frustrations. I believe it is necessary.

**Comment by haileyhoyat** [comment]
@godclause @appills 

Hi All. I want to introduce you to Dan (@SimmonsRitchie ). Dan has officially taken over the role as project lead for the entire City Scrapers project. 

Dan, idk if this conversation is relevant for you, particularly as you fix a lot of infrastructure things. 

Cheers, all.

**Comment by godclause** [comment]
> @godclause @appills @onyangojerry 
> 
> 
> 
> Hi All. I want to introduce you to Dan (@SimmonsRitchie ). Dan has officially taken over the role as project lead for the entire City Scrapers project. 
> 
> 
> 
> Dan, idk if this conversation is relevant for you, particularly as you fix a lot of infrastructure things. 
> 
> 
> 
> Cheers, all. 

Thank you for the introduction @haileyhoyat.  Hi @SimmonsRitchie, nice to meet you here.

**Comment by SimmonsRitchie** [comment]
Hi there, @godclause! Nice to meet you too! And thanks for the intro, @haileyhoyat.

@godclause My apologies, I took over the city-scrapers project very recently and I'm juggling a lot of fixes and upgrades right now across the project's 15 repos. I overlooked this issue and conversation.

**Re: saving PDFs as files before parsing**
I think @appills may have already answered your question, but yes, you can just parse the in-memory byte sequence of the PDF rather than writing it to a file and then parsing it. This is generally more efficient.

**Re: shell/OS issues**
I am a Mac user but I have experienced my own headaches with a number of the city-scraper repos. To my mind I think it may make a lot of sense to dockerize all the projects. I hope this will make them OS-agonistic and improve the dev experience overall (especially for newcomers). I'd be very interested in any feedback on this subject though. Let me know if you have thoughts!

**Comment by godclause** [comment]
> Hi there, @godclause! Nice to meet you too! And thanks for the intro, @haileyhoyat.
> 
> @godclause My apologies, I took over the city-scrapers project very recently and I'm juggling a lot of fixes and upgrades right now across the project's 15 repos. I overlooked this issue and conversation.
> 
> **Re: saving PDFs as files before parsing** I think @appills may have already answered your question, but yes, you can just parse the in-memory byte sequence of the PDF rather than writing it to a file and then parsing it. This is generally more efficient.
> 
> **Re: shell/OS issues** I am a Mac user but I have experienced my own headaches with a number of the city-scraper repos. To my mind I think it may make a lot of sense to dockerize all the projects. I hope this will make them OS-agonistic and improve the dev experience overall (especially for newcomers). I'd be very interested in any feedback on this subject though. Let me know if you have thoughts!

@SimmonsRitchie Hi!

I have 'some' thoughts...

1.  For my parsing issue, @appills's answer did result in a initial inquiry into a need for clarity about OS (mac) updates and how those are affecting python depedencies.

2. How does Docker compare / contrast in 'usability' versus Vagrant for platform (OS) agnosticism?  What's the expected long-term benefit attributed to Docker for City Scrapers projects versus Vagrant, from a support perspective?   I mean, I'm all for improved performance, a better experience for newcomers, etc., but what will implementing Docker instead of Vagrant and vice versa cost City Scrapers' repos?

I'm hoping I'm within scope on these concerns.

**Comment by appills** [comment]
Python module/package dependencies should work regardless of platform, are
you having problems?

**Comment by godclause** [comment]
> > 1. Is the general consensus around scraping PDF documents associated to an idea that they absolutely must be downloaded, stored and locally scraped?
> 
> No need to store them locally. You're going to get a stream of bytes regardless, so you can just read the byte stream directly from the request. I don't know how package management is applied here (e.g. pypdf)
> 
> ```
> from io import BytesIO
> from pypdf import PdfReader
> 
> resp = requests.get("https://hfsrb.illinois.gov/content/dam/soi/en/web/hfsrb/events/documents/2024/january-23-2024-state-board-meeting-/21-007(4)%20Permit%20Renewal%20Winchester%20ASTC.pdf")
> print(resp.headers)
> reader = PdfReader(BytesIO(resp.content))
> print(reader.pages[0].extract_text())
> ```

@appills @onyangojerry 

Hello:

It seems the expected behavior on this code snippet is to parse text from only a single file.  

How does our spider parse pdf's for all future / additional meetings, considering start URL as 'https://www2.illinois.gov/sites/hfsrb/events/Pages/Board-Meetings.aspx'?

**Comment by Rishabhy190** [comment]
I'd like to help

**Comment by godclause** [comment]
@haileyhoyat @SimmonsRitchie 

Hi!  I did read Github documentation for claiming an issue.  It seems end-users need to have Triage access to update labels.  I do not have this access.  Can either of you update the label on my behalf, or provide access so that I can do so?

Also, is there not a slack channel the City Scapers?  I know there was one at the outset.

**Comment by godclause** [comment]
> Python module/package dependencies should work regardless of platform, are
> you having problems?
> […](#)

Hi @appills,

Through careful review, I did learn that Python package dependencies are not platform agnostic, although in an ideal world they should be.  That is the reason for my initial query. 

As far as my thoughts on Docker are concerned, dockerizing each spider can and possibly will future proof the city-scrapers project, long term.  Virtualization could be equally affective, but is likely a bit more complex and more labor intensive.  Virtualization is probably more secure against vulnerabilities that present due to lack of consistent or periodic infrastructure and / or dependency upgrades.  I’m wondering if it would be beneficial to implement a Python / city-scrapers specific notification triggered to generate a notice when an upgrade is available?
